File: frontend/Features/Profile/PaywallView.swift
import SwiftUI
import Foundation

// MARK: - 结账弹层（原型 prof-modal · §10 demo mock 支付）
// 解锁完整经验 ¥9.9 / 咨询 ¥29 / 资料包 / 陪跑 ¥599 —— demo 直接写 unlocks，不接真 IAP。

struct PaywallView: View {
    @Bindable var model: ProfileModel
    let checkout: ProfileModel.Checkout

    @Environment(SupabaseService.self) private var supabase
    @State private var processing = false
    @State private var succeeded = false
    /// 付费解锁是关键动作：游客先登录（AuthGate 就地弹 LoginSheet）
    @State private var authGate = AuthGateCenter()

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                if succeeded { successView } else { form }
            }
            .padding(.horizontal, 22).padding(.top, 12).padding(.bottom, 34)
        }
        .scrollIndicators(.hidden)
        .authGate(authGate)
    }

    // MARK: 结账表单

    private var form: some View {
        VStack(alignment: .leading, spacing: 0) {
            HStack {
                Text(headTitle).font(.system(size: 18, weight: .semibold)).foregroundStyle(Theme.ink)
                Spacer()
                Button { model.checkout = nil } label: {
                    Image(systemName: "xmark").font(.system(size: 13, weight: .semibold)).foregroundStyle(Theme.sub)
                        .frame(width: 32, height: 32).background(Theme.raised, in: Circle())
                }
                .buttonStyle(.plain)
            }

            orderCard.padding(.top, 18)

            Text(payNote).font(.system(size: 11)).foregroundStyle(Theme.faint).lineSpacing(5).padding(.top, 13)

            Button(action: pay) {
                Group {
                    if processing {
                        ProgressView().tint(.white)
                    } else {
                        Text(actionTitle).font(.system(size: 14, weight: .semibold))
                    }
                }
                .foregroundStyle(.white)
                .frame(maxWidth: .infinity).padding(.vertical, 15)
                .background(Theme.buttonGradient, in: Capsule())
            }
            .buttonStyle(PressScaleStyle())
            .disabled(processing)
            .padding(.top, 20)
        }
    }

    private var orderCard: some View {
        HStack(spacing: 14) {
            VStack(alignment: .leading, spacing: 4) {
                Text(itemTitle).font(.system(size: 13.5, weight: .semibold)).foregroundStyle(Theme.ink)
                Text(itemSub).font(.system(size: 11)).foregroundStyle(Theme.sub)
                    .frame(maxWidth: .infinity, alignment: .leading)
            }
            Text("¥\(priceText)").font(.system(size: 22, weight: .bold)).foregroundStyle(Theme.ink)
                .fixedSize()
        }
        .padding(16)
        .background(Theme.card, in: RoundedRectangle(cornerRadius: 18, style: .continuous))
        .overlay(RoundedRectangle(cornerRadius: 18, style: .continuous).strokeBorder(Theme.line, lineWidth: 1))
    }

    // MARK: 成功态

    private var successView: some View {
        VStack(spacing: 0) {
            Image(systemName: "checkmark").font(.system(size: 28, weight: .bold)).foregroundStyle(Theme.teal)
                .frame(width: 62, height: 62).background(Color(hex: 0x3ED9A4, alpha: 0.14), in: Circle())
            Text(successTitle).font(.system(size: 18, weight: .semibold)).foregroundStyle(Theme.ink).padding(.top, 16)
            Text(successSub).font(.system(size: 12)).foregroundStyle(Theme.sub).lineSpacing(5)
                .multilineTextAlignment(.center).padding(.top, 8)
            PrimaryButton(title: "好的") { model.checkout = nil }.padding(.top, 24)
        }
        .frame(maxWidth: .infinity)
        .padding(.top, 20)
    }

    // MARK: 动作

    private func pay() {
        authGate.require(supabase) { confirmPay() }
    }

    private func confirmPay() {
        processing = true
        Task {
            if case .unlock = checkout {
                await model.confirmUnlock(supabase: supabase)
            } else {
                try? await Task.sleep(for: .milliseconds(600))   // demo：模拟下单
            }
            processing = false
            withAnimation(.spring(response: 0.4, dampingFraction: 0.8)) { succeeded = true }
        }
    }

    // MARK: 文案

    private var headTitle: String {
        switch checkout {
        case .unlock: "解锁完整经验"
        case .service(let s): s.kind == "consult" ? "预约咨询" : "确认订单"
        }
    }
    private var actionTitle: String {
        switch checkout {
        case .unlock: "确认解锁 · ¥\(priceText)"
        case .service(let s): s.kind == "consult" ? "确认预约 · ¥\(priceText)" : "确认购买 · ¥\(priceText)"
        }
    }
    private var itemTitle: String {
        switch checkout {
        case .unlock: "\(model.traveler?.name ?? "TA") · 完整转型经验"
        case .service(let s): s.title
        }
    }
    private var itemSub: String {
        switch checkout {
        case .unlock: "完整故事 + 全部踩坑建议 + 完整轨迹"
        case .service(let s): s.description
        }
    }
    private var priceText: String {
        switch checkout {
        case .unlock: NSDecimalNumber(decimal: AppConfig.Price.unlockProfile).stringValue
        case .service(let s): NSDecimalNumber(decimal: s.price).stringValue
        }
    }
    private let payNote = "演示环境：点击即模拟完成，不产生真实扣款。正式版将通过 App 内购（StoreKit 2）安全支付。"

    private var successTitle: String {
        switch checkout {
        case .unlock: "已解锁完整经验"
        case .service(let s): s.kind == "consult" ? "预约成功" : "购买成功"
        }
    }
    private var successSub: String {
        switch checkout {
        case .unlock: "完整故事、踩坑建议与轨迹已全部展开，回到主页查看。"
        case .service: "TA 会尽快与你确认。可在消息中追问具体细节。"
        }
    }
}

File: android/app/src/main/java/app/possibility/android/features/profile/PaywallView.kt
package app.possibility.android.features.profile

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.possibility.android.core.AppConfig
import app.possibility.android.core.model.Traveler
import app.possibility.android.core.model.TravelerServiceItem
import app.possibility.android.core.network.SupabaseService
import app.possibility.android.core.theme.Theme
import app.possibility.android.features.auth.AuthGateCenter
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

// 结账弹层（原型 prof-modal · §10 demo mock 支付）
// 对应 ios/Possibility/Features/Profile/PaywallView.swift：单商品订单卡，
// 由入口决定结账对象 —— 锁定块 → 解锁完整经验；服务卡 → 该项服务。
// 解锁走真实 unlocks 写入；具体服务（咨询 / 资料包 / 陪跑）demo 模拟下单。

/** 结账类型，对应 iOS ProfileModel.Checkout。 */
sealed interface ProfileCheckout {
    data object Unlock : ProfileCheckout
    data class Service(val item: TravelerServiceItem) : ProfileCheckout
}

private const val PAY_NOTE =
    "演示环境：点击即模拟完成，不产生真实扣款。正式版将通过 App 内购安全支付。"

/** 付费墙底部弹层：单商品订单卡 + mock 支付。付费是关键动作，先过登录门控。 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun PaywallSheet(
    traveler: Traveler,
    checkout: ProfileCheckout,
    onDismiss: () -> Unit,
    onUnlocked: () -> Unit,
) {
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val scope = rememberCoroutineScope()

    var processing by remember { mutableStateOf(false) }
    var succeeded by remember { mutableStateOf(false) }

    val service = (checkout as? ProfileCheckout.Service)?.item
    val consult = service?.kind == "consult"

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = Color(0xFF11141D),
        dragHandle = null,
    ) {
        Column(
            Modifier
                .fillMaxWidth()
                .navigationBarsPadding()
                .padding(horizontal = 22.dp)
                .padding(top = 18.dp, bottom = 28.dp),
        ) {
            if (succeeded) {
                SuccessView(
                    title = when {
                        service == null -> "已解锁完整经验"
                        consult -> "预约成功"
                        else -> "购买成功"
                    },
                    subtitle = if (service == null) {
                        "完整故事、踩坑建议与轨迹已全部展开，回到主页查看。"
                    } else {
                        "TA 会尽快与你确认。可在消息中追问具体细节。"
                    },
                    onDone = onDismiss,
                )
            } else {
                // 标题栏
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        when {
                            service == null -> "解锁完整经验"
                            consult -> "预约咨询"
                            else -> "确认订单"
                        },
                        color = Theme.ink,
                        fontSize = 18.sp,
                        fontWeight = FontWeight.SemiBold,
                        modifier = Modifier.weight(1f),
                    )
                    Box(
                        Modifier
                            .size(32.dp)
                            .clip(CircleShape)
                            .background(Theme.raised)
                            .clickableNoRipple(onDismiss),
                        contentAlignment = Alignment.Center,
                    ) {
                        Text("✕", color = Theme.sub, fontSize = 13.sp, fontWeight = FontWeight.SemiBold)
                    }
                }

                // 订单卡（iOS orderCard）
                val price = service?.price ?: AppConfig.Price.UNLOCK_PROFILE
                Row(
                    Modifier
                        .padding(top = 18.dp)
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(18.dp))
                        .background(Theme.card)
                        .border(1.dp, Theme.line, RoundedCornerShape(18.dp))
                        .padding(16.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Column(Modifier.weight(1f)) {
                        Text(
                            service?.title ?: "${traveler.name} · 完整转型经验",
                            color = Theme.ink,
                            fontSize = 13.5.sp,
                            fontWeight = FontWeight.SemiBold,
                        )
                        Text(
                            service?.description ?: "完整故事 + 全部踩坑建议 + 完整轨迹",
                            color = Theme.sub,
                            fontSize = 11.sp,
                            lineHeight = 16.sp,
                            modifier = Modifier.padding(top = 4.dp),
                        )
                    }
                    Spacer(Modifier.width(14.dp))
                    Text("¥${formatPrice(price)}", color = Theme.ink, fontSize = 22.sp, fontWeight = FontWeight.Bold)
                }

                Text(
                    PAY_NOTE,
                    color = Theme.faint,
                    fontSize = 11.sp,
                    lineHeight = 16.sp,
                    modifier = Modifier.padding(top = 13.dp),
                )

                Box(
                    Modifier
                        .padding(top = 20.dp)
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(999.dp))
                        .background(Theme.buttonGradient)
                        .clickableNoRipple {
                            if (processing) return@clickableNoRipple
                            // 与 iOS pay() 对齐：付费是关键动作，游客先被登录门控拦下，
                            // 登录成功后继续本次支付（AuthGateHost 由旅人主页根部挂载）。
                            AuthGateCenter.require("paywall") {
                                processing = true
                                scope.launch {
                                    if (service == null) {
                                        // 与 iOS confirmUnlock 一致：只有写库成功才标记解锁，
                                        // 但 demo 弹层无论成败都进成功态。
                                        if (SupabaseService.shared.unlockProfile(traveler.id)) onUnlocked()
                                    } else {
                                        delay(600)   // demo：模拟下单
                                    }
                                    processing = false
                                    succeeded = true
                                }
                            }
                        }
                        .padding(vertical = 15.dp),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        when {
                            processing -> "处理中…"
                            service == null -> "确认解锁 · ¥${formatPrice(price)}"
                            consult -> "确认预约 · ¥${formatPrice(price)}"
                            else -> "确认购买 · ¥${formatPrice(price)}"
                        },
                        color = Color.White,
                        fontSize = 14.sp,
                        fontWeight = FontWeight.SemiBold,
                    )
                }
            }
        }
    }
}

@Composable
private fun SuccessView(title: String, subtitle: String, onDone: () -> Unit) {
    Column(
        Modifier
            .fillMaxWidth()
            .padding(top = 20.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Box(
            Modifier
                .size(62.dp)
                .clip(CircleShape)
                .background(Theme.teal.copy(alpha = 0.14f)),
            contentAlignment = Alignment.Center,
        ) {
            Icon(Icons.Filled.Check, contentDescription = null, tint = Theme.teal, modifier = Modifier.size(28.dp))
        }
        Text(
            title,
            color = Theme.ink,
            fontSize = 18.sp,
            fontWeight = FontWeight.SemiBold,
            modifier = Modifier.padding(top = 16.dp),
        )
        Text(
            subtitle,
            color = Theme.sub,
            fontSize = 12.sp,
            lineHeight = 18.sp,
            modifier = Modifier.padding(top = 8.dp),
        )
        Box(
            Modifier
                .padding(top = 24.dp)
                .fillMaxWidth()
                .clip(RoundedCornerShape(999.dp))
                .background(Theme.buttonGradient)
                .clickableNoRipple(onDone)
                .padding(vertical = 15.dp),
            contentAlignment = Alignment.Center,
        ) {
            Text("好的", color = Color.White, fontSize = 14.sp, fontWeight = FontWeight.SemiBold)
        }
    }
}
